### PR TITLE
Port Custom Resource and Spark Operator provisioners from EG

### PR DESCRIPTION
**Pull request by kevin-bates**
This pull request ports EG's `CustomResourceProcessProxy` and `SparkOperatorProcessProxy` to the functionally equivalent `CustomResourceProvisioner` and `SparkOperatorProvisioner`, respectively.  Also included in these changes are the corresponding doc updates and a new CLI option for `jupyter k8s-spec install` of `--crd`. 

Some additional changes related to recent bug fixes were also included in the port.

These changes cover the following EG PRs (all of which are labeled with `gateway-provisioners`):

https://github.com/jupyter-server/enterprise_gateway/pull/991
https://github.com/jupyter-server/enterprise_gateway/pull/1014
https://github.com/jupyter-server/enterprise_gateway/pull/1031
https://github.com/jupyter-server/enterprise_gateway/pull/1087
https://github.com/jupyter-server/enterprise_gateway/pull/1164
https://github.com/jupyter-server/enterprise_gateway/pull/1175
https://github.com/jupyter-server/enterprise_gateway/pull/1249
https://github.com/jupyter-server/enterprise_gateway/pull/1260
https://github.com/jupyter-server/enterprise_gateway/pull/1267
https://github.com/jupyter-server/enterprise_gateway/pull/1269
https://github.com/jupyter-server/enterprise_gateway/pull/1271

Special thanks to the following who contributed to the original changes in Enterprise Gateway and will be added as co-authors on the merge:
@abzymeinsjtu, @tahesse, @lresende, @kiersten-stokes, @manoj-jeswani, @pq43, @rahul26goyal, @amarinder-bindra

Resolves: #37

### CODE REVIEW

**Comment by kevin-bates** [comment]
This PR is ready for review.  It is in Draft mode for another day until I can confirm (or not) the co-authors (see #37).

I'm hoping any of @echarles, @kiersten-stokes, or @lresende can take this branch for a spin.   If not a "spin" perhaps a visual scan. :eyes:

I found the best approach is to use the instructions in the [Application Support README](https://github.com/jupyter-server/gateway_provisioners/blob/main/gateway_provisioners/app-support/README.md) to:

1.  Build a JKG image containing the default kernelspecs along with the python kernel images (you can build kernel images using `make kernel-images`):
```bash
make dist gp-spark-base gp-kernel-py gp-kernel-spark-py gp-jkg
```
2. Deploy JKG into Kubernetes using the helm command listed in the app support doc:
```bash
helm  upgrade --install  gateway-provisioners gateway_provisioners/app-support/kubernetes/helm/gateway-provisioners -n gateway-provisioners
```
(Note: namespace `gateway-provisioners` must be created prior to `helm install`: `kubectl create namespace gateway-provisioners`.)
3. Confirm startup by looking at the logs: 
```bash
kubectl logs -f deployment.apps/gateway-provisioners -n gateway-provisioners
```
You should see the startup banner for JKG. 
4. You'll need to expose the port on the Gateway Provisioners service.  I use the port-forwarding feature in Rancher Desktop and expose port `53434`.
5. Then launch lab, pointing the gateway-url at your JKG deployment:
```bash
jupyter lab --debug --notebook-dir=~/notebooks --gateway-url=http://localhost:53434
```
6. Then launch the kernel named "Kubernetes Spark Operator".
Here's some code to exercise the kernel:

```python
# Create a spark context...

from pyspark.sql import SparkSession

sc = SparkSession.builder.getOrCreate().sparkContext
sc
```
```python
# Print the version...

sc.version
```
```python
# Run some spark code...

import time

n = 1000000
start=time.time()
total=sc.parallelize(range(n)).sum()
end=time.time()
print(f"Sum of 0..{n-1} is {total} in {end-start:.3f}s")
```
```python
# Test interrupt by interrupting the sleep call...
import time
print('begin')
time.sleep(60)
print('end')
```
```python
# Test restart by entering `a = 123` in one cell.  And referencing `a` in a second cell, 
# but issue a restart between those cells.  Upon the completion of the restart, execute 
# the second cell and ensure a "not defined" exception is returned.

a = 123

# issue restart and execute the following (and ensure error):

a
```

To remove your deployment, the following can be issued:
```bash
helm delete enterprise-gateway -n enterprise-gateway
```

**Comment by echarles** [comment]
This is the result of my test with your branch on my local minikube.

```bash
eval $(minikube docker-env)
make dist gp-spark-base gp-kernel-py gp-kernel-spark-py gp-jkg
helm  upgrade --install  gateway-provisioners gateway_provisioners/app-support/kubernetes/helm/gateway-provisioners -n gateway-provisioners --create-namespace
kubectl logs -f deployment.apps/gateway-provisioners -n gateway-provisioners
# [D 2023-04-20 05:00:49.466 KernelGatewayApp] Found kernel k8s_python_spark_operator in /usr/local/share/jupyter/kernels
# [I 230420 05:00:49 web:2327] 200 GET /api/kernelspecs (10.244.0.43) 15.61ms
minikube service gateway-provisioners --url -n gateway-provisioners
# http://192.168.64.38:31470
# http://192.168.64.38:30952
jupyter lab --notebook-dir=~/notebooks --gateway-url=http://192.168.64.38:31470
# 1. Launch the kernel named "Kubernetes Python".
# -- Works fine, jovyan-4123d721-1b38-493d-9db3-bc42ebc18e64 is launched.
# 2. Launch the kernel named "Kubernetes Spark Operator".
# -- Fails with tornado.web.HTTPError: HTTP 500: Internal Server Error (Error from Gateway: [Unknown server error] Internal Server Error. Ensure gateway url is valid and the Gateway instance is running.)
# -- No error shown in kubectl logs -f deployment.apps/gateway-provisioners -n gateway-provisioners
```

**Comment by kevin-bates** [comment]
Hi @echarles - first of all, thank you very much for heading down this road.  As I was reading your description (thank you for checking the vanilla k8s kernel) it dawned on me that I forgot to mention that the SparkOperator K8s object requires installation.  This is described in [this portion](https://github.com/jupyter-server/gateway_provisioners/pull/80/files#diff-f397c0b6ff9a1beb29c7736292ff3c2076319bb4f736bee1dc3173b75863bee0R109-R119) of this PR, but, as the instructions are identical, you can also reference that same "Note snippet" in the EG docs [here](https://jupyter-enterprise-gateway.readthedocs.io/en/latest/operators/deploy-kubernetes.html#overview).

Prior to doing that, it would be good to take a look at the Gateway Provisioner/JKG logs using:
```bash
kubectl logs deployment.apps/gateway-provisioners -n gateway-provisioners
```
Following the launch failure, there should be some mention that the Kubernetes Spark Operator is not installed.  Once that is installed, the kernel should launch.  You should not need to redeploy GP/JKG.  A retry of the kernel's creation (following the operator's installation) should then succeed.

Thank you!

**Comment by echarles** [comment]
>  it dawned on me that I forgot to mention that the SparkOperator K8s object requires installation.

thx, I have run that command to install the spark-operator

```
bash
helm upgrade --install spark-operator spark-operator/spark-operator --namespace spark-operator --set webhook.enable=true  --create-namespace
```

> Following the launch failure, there should be some mention that the Kubernetes Spark Operator is not installed

Oops, forgot to check...

>  Once that is installed, the kernel should launch. You should not need to redeploy GP/JKG. A retry of the kernel's creation (following the operator's installation) should then succeed.

Yeah, after the spark-operator installation, without reinstalling anything else, it now works. Great job Kevin!

Screenshot 2023-04-20 at 16 23 22

PS: kubectl logs -f deployment.apps/gateway-provisioners -n gateway-provisioners constantly generates the following after the spark-operator installation. Something needs to be fixed there...

```
[I 230420 05:20:41 web:2327] 200 GET /api/kernelspecs (10.244.0.43) 18.50ms
[D 2023-04-20 05:20:46.725 KernelGatewayApp] Found kernel python3 in /opt/conda/share/jupyter/kernels
[D 2023-04-20 05:20:46.727 KernelGatewayApp] Found kernel k8s_python in /usr/local/share/jupyter/kernels
[D 2023-04-20 05:20:46.728 KernelGatewayApp] Found kernel k8s_scala_spark in /usr/local/share/jupyter/kernels
[D 2023-04-20 05:20:46.728 KernelGatewayApp] Found kernel docker_scala in /usr/local/share/jupyter/kernels
[D 2023-04-20 05:20:46.728 KernelGatewayApp] Found kernel docker_r in /usr/local/share/jupyter/kernels
[D 2023-04-20 05:20:46.728 KernelGatewayApp] Found kernel docker_swarm_scala in /usr/local/share/jupyter/kernels
[D 2023-04-20 05:20:46.728 KernelGatewayApp] Found kernel docker_swarm_python in /usr/local/share/jupyter/kernels
[D 2023-04-20 05:20:46.728 KernelGatewayApp] Found kernel k8s_r_spark in /usr/local/share/jupyter/kernels
[D 2023-04-20 05:20:46.728 KernelGatewayApp] Found kernel docker_swarm_r in /usr/local/share/jupyter/kernels
[D 2023-04-20 05:20:46.728 KernelGatewayApp] Found kernel k8s_r in /usr/local/share/jupyter/kernels
[D 2023-04-20 05:20:46.728 KernelGatewayApp] Found kernel docker_python in /usr/local/share/jupyter/kernels
[D 2023-04-20 05:20:46.728 KernelGatewayApp] Found kernel k8s_scala in /usr/local/share/jupyter/kernels
[D 2023-04-20 05:20:46.728 KernelGatewayApp] Found kernel k8s_python_spark in /usr/local/share/jupyter/kernels
[D 2023-04-20 05:20:46.729 KernelGatewayApp] Found kernel k8s_python_spark_operator in /usr/local/share/jupyter/kernels
[I 230420 05:20:46 web:2327] 200 GET /api/kernelspecs (10.244.0.43) 17.24ms
[D 2023-04-20 05:20:51.760 KernelGatewayApp] Found kernel python3 in /opt/conda/share/jupyter/kernels
[D 2023-04-20 05:20:51.761 KernelGatewayApp] Found kernel k8s_python in /usr/local/share/jupyter/kernels
[D 2023-04-20 05:20:51.762 KernelGatewayApp] Found kernel k8s_scala_spark in /usr/local/share/jupyter/kernels
[D 2023-04-20 05:20:51.762 KernelGatewayApp] Found kernel docker_scala in /usr/local/share/jupyter/kernels
[D 2023-04-20 05:20:51.762 KernelGatewayApp] Found kernel docker_r in /usr/local/share/jupyter/kernels
[D 2023-04-20 05:20:51.763 KernelGatewayApp] Found kernel docker_swarm_scala in /usr/local/share/jupyter/kernels
[D 2023-04-20 05:20:51.763 KernelGatewayApp] Found kernel docker_swarm_python in /usr/local/share/jupyter/kernels
[D 2023-04-20 05:20:51.763 KernelGatewayApp] Found kernel k8s_r_spark in /usr/local/share/jupyter/kernels
[D 2023-04-20 05:20:51.763 KernelGatewayApp] Found kernel docker_swarm_r in /usr/local/share/jupyter/kernels
...
```

**Comment by kevin-bates** [comment]
Thank you Eric.  It definitely takes a village! :smile:

I'm going to remove this from draft.  If you could go ahead and peruse the changes and submit a review, we can then move this forward with a 0.2.0 release.

Thanks again!

File: docs/source/contributors/system-architecture.md [hunk 1]
```diff
@@ -61,44 +64,41 @@ blockdiag {
...
         """
         Checks to see if the kernel launch timeout has been exceeded while awaiting connection info.
         """
 ```
 
-Kernel launch timeout expiration is expressed via the environment variable `KERNEL_LAUNCH_TIMEOUT`. If this
-value does not exist, it defaults to the host application process environment variable `KERNEL_LAUNCH_TIMEOUT` - which
-defaults to 30 seconds if unspecified. Since all `KERNEL_` environment variables "flow" from the Notebook server, the launch
-timeout can be specified as a client attribute of the Notebook session.
-
```
Comment (echarles) [inline docs/source/contributors/system-architecture.md:100]: This PR removes information related to KERNEL_LAUNCH_TIMEOUT. Is this intended or just a carry-on opportunity?

Comment (kevin-bates) [reply docs/source/contributors/system-architecture.md:100]: yes, this was intentional.  That paragraph just didn't flow and the equivalent information is already defined in [Kernel Environment Variables](file:///Users/kbates/repos/oss/jupyter/gateway_provisioners/docs/build/html/users/kernel-envs.html#kernel-environment-variables) in the Users guide.

**Review by echarles — APPROVED** [review]
LGTM. Thx @kevin-bates

**Pull request merged**